### PR TITLE
Promotion claim date

### PR DESCRIPTION
**Pull request by NejcZdovc**
Resolves https://github.com/brave/brave-browser/issues/7297

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
covered with unit tests

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_rewards/browser/database/database_promotion.cc [hunk 1]
```diff
@@ -12,6 +12,11 @@
 #include "sql/statement.h"
 #include "sql/transaction.h"
 
+namespace {
+  const char* table_name_ = "promotion";
```
Comment (tmancey) [inline components/brave_rewards/browser/database/database_promotion.cc]: nit: `const char[] table_name = "promotion"`

File: components/brave_rewards/browser/database/database_promotion.cc [hunk 2]
```diff
@@ -12,6 +12,11 @@
...
+  const int minimum_version_ = 10;
```
Comment (tmancey) [inline components/brave_rewards/browser/database/database_promotion.cc]: not a member variable so does not require the underscore

File: components/brave_rewards/browser/database/database_promotion.cc (hunk 1 above)
Comment (tmancey) [reply components/brave_rewards/browser/database/database_promotion.cc]: not a member variable so does not require the underscore

File: components/brave_rewards/browser/database/database_promotion.cc [hunk 3]
```diff
@@ -76,10 +118,69 @@ bool DatabasePromotion::CreateTable(sql::Database* db) {
...
+    }
+  }
+}
+
+bool DatabasePromotion::MigrateToV10(sql::Database* db) {
+  if (!CreateTableV10(db)) {
+    return false;
+  }
+
+  if (!CreateTableV10(db)) {
```
Comment (tmancey) [inline components/brave_rewards/browser/database/database_promotion.cc]: duplicated code?

File: components/brave_rewards/browser/database/database_promotion.cc [hunk 4]
```diff
@@ -76,10 +118,69 @@ bool DatabasePromotion::CreateTable(sql::Database* db) {
...
+}
+
+bool DatabasePromotion::MigrateToV13(sql::Database* db) {
+  if (!db->DoesTableExist(table_name_)) {
+    if (!CreateTableV13(db)) {
+      return false;
+    }
+  }
+
+  const char* column = "claimed_at";
```
Comment (tmancey) [inline components/brave_rewards/browser/database/database_promotion.cc]: nit: `const char[]`

File: components/brave_rewards/browser/database/publisher_info_database.cc [hunk 5]
```diff
@@ -796,6 +809,9 @@ bool PublisherInfoDatabase::Migrate(int version) {
     case 12: {
       return MigrateV11toV12();
     }
+    case 13: {
+      return MigrateV12toV13();
+    }
     default:
       return false;
```
Comment (tmancey) [inline components/brave_rewards/browser/database/publisher_info_database.cc]: NOTREACHED()?

File: components/brave_rewards/browser/database/publisher_info_database_unittest.cc [hunk 6]
```diff
@@ -1293,6 +1293,26 @@ TEST_F(PublisherInfoDatabaseTest, Migrationv11tov12_ContributionInfo) {
   EXPECT_EQ(list.at(3)->publisher_key, "reddit.com");
 }
 
+TEST_F(PublisherInfoDatabaseTest, Migrationv12tov13) {
+  base::ScopedTempDir temp_dir;
+  base::FilePath db_file;
+  CreateMigrationDatabase(&temp_dir, &db_file, 12, 13);
+  EXPECT_TRUE(publisher_info_database_->Init());
```
Comment (tmancey) [inline components/brave_rewards/browser/database/publisher_info_database_unittest.cc]: `EXPECT_TRUE` yields a fatal error, is this expected over `ASSERT_TRUE` which yields a non-fatal error

Comment (NejcZdovc) [reply components/brave_rewards/browser/database/publisher_info_database_unittest.cc]: it's vice-versa, ASSERT is fatal, EXPECT is non-fatal

Comment (tmancey) [reply components/brave_rewards/browser/database/publisher_info_database_unittest.cc]: yeap I've got things back to front...

File: components/brave_rewards/browser/database/database_promotion.cc [hunk 7]
```diff
@@ -76,10 +118,69 @@ bool DatabasePromotion::CreateTable(sql::Database* db) {
...
+  if (!db->DoesColumnExist(table_name_, column)) {
+    const std::string query = base::StringPrintf(
+        "ALTER TABLE %s ADD %s TIMESTAMP;",
```
Comment (gdregalo) [inline components/brave_rewards/browser/database/database_promotion.cc]: do you need the semicolon in the end of sql statement?

Comment (NejcZdovc) [reply components/brave_rewards/browser/database/database_promotion.cc]: not needed, can remove

File: components/brave_rewards/browser/database/database_promotion.cc (hunk 3 above)
Comment (NejcZdovc) [reply components/brave_rewards/browser/database/database_promotion.cc]: yup, should be index. Fixed

File: components/brave_rewards/browser/database/database_promotion.cc (hunk 1 above)
Comment (NejcZdovc) [reply components/brave_rewards/browser/database/database_promotion.cc]: covered in https://github.com/brave/brave-browser/issues/7145. This code was not added it was just moved from h file

File: components/brave_rewards/browser/database/database_promotion.cc (hunk 2 above)
Comment (NejcZdovc) [reply components/brave_rewards/browser/database/database_promotion.cc]: covered in https://github.com/brave/brave-browser/issues/7145. This code was not added it was just moved from h file

File: components/brave_rewards/browser/database/database_promotion.cc (hunk 4 above)
Comment (NejcZdovc) [reply components/brave_rewards/browser/database/database_promotion.cc]: done

File: components/brave_rewards/browser/database/publisher_info_database.cc (hunk 5 above)
Comment (NejcZdovc) [reply components/brave_rewards/browser/database/publisher_info_database.cc]: added

File: components/brave_rewards/browser/database/database_promotion.cc (hunk 7 above)
Comment (NejcZdovc) [reply components/brave_rewards/browser/database/database_promotion.cc]: removed

**Comment by NejcZdovc** [comment]
CI passed on all OS except on windows (build didn't start). Restarting windows

**Pull request merged**